### ISSUE TITLE
Can't run all tests in sub group, if root group is skipped.

### ISSUE THREAD

**Issue by MiniSuperDev** [Dart-Code/Dart-Code#4503]
Hello, I have root group test with skip value, for example when tests some rest api points, this is only for dev.
Inside I have some sub groups, that I want to execute from the UI, but all tests in this subgroup are skipped, and I'll be forced to execute only tests and not subgroups.

Thank you

```dart
import 'package:test/test.dart';

void main() {
  group('skipped root group', () {
    group(
      'sub group',
      () {
        test('failing test', () {
          fail('I am a failure');
        });
      },
    );
  }, skip: 'SKIPPED FROM ROOT GROUP');
}
```

```shell
Skip: SKIPPED FROM ROOT GROUP
✓ skipped root group sub group failing test
```

![image](https://user-images.githubusercontent.com/105388540/233475659-52a09509-f253-4a75-995d-5c4d9255814d.png)

**Comment by DanTup** [comment]
Ok, I discovered why we have this behaviour...

When you run a test explicitly using the CodeLens links, we pass `--run-skipped` to `package:test` to say that we want to run the test regardless of it's skip status (to allow you to easily run skipped tests).

We don't do this with groups, because passing `--run-skipped` will run all skipped tests within the group, not only the group that was skipped. For example:

```dart
  group('group', () {
    test('failing test', () {
      fail('I am a failure');
    });
    test('skipped test', () {
      fail('I am a failure');
    }, skip: 'SKIPPED TEST');
  });
```

Here, a test within the group is skipped. If we pass the `--run-skipped` flag for the group, then the skipped test will be run - but probably that's not what you expected.

I think the desired behaviour when running a group is to say "run the group even if it is skipped, but honour the skipped flag on any _children_".

@jakemac53 any thoughts on this? It's probably a bit odd because were not really "running the group" as far as package:test is concerned, I think we're just running "all the tests that start with `"skipped group "` so distinguishing between "the group being run" and "children of the group being run" is probably a bit less obvious. Although maybe that would make more sense when running by line? If we said "run the thing on line 5 with --run-skipped", would it be more reasonable to only ignore skipped for the thing on line 5 and honour skip flags on children?

**Comment by jakemac53** [comment]
I would have to look at how feasible it is, but what if we just always ignored any `skip` configuration if running a test by line and column (but only ignore it for the exact thing at that line/column, whether it is a group or test)? Any child configuration would still be respected.

That seems like generally a safe thing to do, as without `--run-skipped` it would just do nothing at all, but you very explicitly asked to run exactly that thing (without ambiguity like with the name matching).

**Comment by DanTup** [comment]
That sounds reasonable to me, however in VS Code we might sometimes run with line numbers and sometimes run with names (the user can choose a preference, but there may be cases where we end up with one and not the other), so this could lead to inconsistent results.

Having a way to run a group by name could also allow you to do it for names, but today we don't have a way to do that:


This doesn't work, because it only matches test names, and the `$` prevents matching the tests inside it:

```
dart test --name "^my group name$"
```

Dropping the `$` means we now could match more than we expected. For example this runs "group one hundred" too:

```
dart test --name "^group one"
```

If we could do something like:

```
dart test --group-name "^group one$"
dart test --group --name "^group one$"
```

That would allow us to support this consistently in both modes? (eg. `--run-skipped` only applies to the thing matching the query, and not children that were implicitly included)

**Comment by jakemac53** [comment]
Hmm, in theory we could do something like that but I am hesitant to further expand the API here, when we already have a feature that should be better for this use case (running by line/col). What are the situations where that doesn't work well?

**Comment by DanTup** [comment]
I'll have to do some testing to get some concrete examples (if I can) - but as an example in the JSON reporter API here, line/column are nullable:

https://github.com/dart-lang/test/blob/master/pkgs/test/doc/json_reporter.md

So if the user runs a test and we don't get a line/col, but the name was also dynamic, we will end up producing a node in the test explorer with the name of the test, but won't know it's line/col (`package:test` didn't give us one, and we failed to match it to an analyzer-discovered test).

I don't know under what situations `package:test` wouldn't give us a line/col though, just that my APIs are also nullable because of this. Do you know what might lead to that? (or is the protocol this way simply because older versions didn't provide it)?

**Comment by jakemac53** [comment]
I believe this would happen primarily when the source information isn't available - for instance if we were running a JS tests and had no sourcemap available. Or a wasm test where source maps aren't supported at all yet.

**Comment by jakemac53** [comment]
cc @natebosch in case you have thoughts here

**Comment by jakemac53** [comment]
Another thought which we keep kicking around here is to just have a more interactive service that you interact with. So you could ask for the available tests, they would have longer lived and consistent IDs, and you could just run by ID.

**Comment by jakemac53** [comment]
@DanTup Given the situations where we don't have a Dart stack trace are pretty restricted (wasm will eventually get support, and we should almost always have a JS source map), what if we started by just doing the line/col thing I suggested above? As long as DartCode prefers to run by line/col when it has them, then this should work for the vast majority of use cases?

**Comment by DanTup** [comment]
@jakemac53 I _think_ that might be reasonable, but I'll try and do some testing with run-by-line enabled tomorrow and log any cases where it's not working. Right now it's hard almost zero testing (besides me and the CI) so I can't say with confidence how well it's working (I've had it turned on myself for a while, but a lot of my work is in the SDK where I can't currently use the test runner 😄).

(I'll post back here in the next day or two with something more concrete)

**Comment by jakemac53** [comment]
Ok, I am also looking into the implementation work that would be required for this and it is actually pretty hard to slot in, but I am sure we could make it work, it just won't be easy.

**Comment by jakemac53** [comment]
Yeah, unfortunately I think my idea just really isn't going to be compatible with how the line/col functionality works.

Basically the problem is that today we select based on line/col if _any_ frame in the trace matches the line/col configuration. This is how running a test nested within a group actually works, we don't ever match the groups at all directly. This also is what enables custom wrapper test functions to work, and I don't think we would want to sacrifice that. But this means we would end up back at the current behavior - any skipped test within the group would end up running.

I do wonder if this is worth really supporting at all though, it does seem like an edge case to me. Longer term if we can somehow support running tests by ID then we could support this better.

**Comment by DanTup** [comment]
Sounds reasonable to me. If one day we might end up with a "service" for this kind of thing, then it does seem less worthwhile changing how this works in the meantime. 

I'll keep this open in the backlog since I think it's a fair request to track, even if it's not something likely to happen in the short-term.

**Comment by DanTup** [comment]
I'm tidying up the issue tracker and closing things that are unlikely (or unable) to be worked on in the foreseeable future. Since there isn't a clear solution to this and it's a minor issue with no 👍's, I'm going to close this.

If it turns out this is a major issue and we can come up with a reasonable solution, we can re-open and revisit this.
